src/annotate.rs

```diff
@@ -37,65 +37,92 @@ pub enum MemberId<'a> {
 pub trait Annotate {
     fn format(&self, variant: Option<&str>, field: &MemberId) -> Option<Format>;
     fn comment(&self, variant: Option<&str>, field: &MemberId) -> Option<String>;
+    fn as_annotate(&self) -> Option<&dyn Annotate>;
+    fn thunk_serialize(&self, serializer: &mut AnnotatedSerializer) -> Result<Document, Error>;
 }
 
-type IdFn = fn() -> usize;
-type CastFn = unsafe fn(*const ()) -> &'static dyn Annotate;
+/// The default implementation of Annotate returns no comments or annotations and
+/// cannot return the trait object.
+impl<T: ?Sized + serde::Serialize> Annotate for T {
+    default fn format(&self, _variant: Option<&str>, _field: &MemberId) -> Option<Format> {
+        None
+    }
+    default fn comment(&self, _variant: Option<&str>, _field: &MemberId) -> Option<String> {
+        None
+    }
+    default fn as_annotate(&self) -> Option<&dyn Annotate> {
+        None
+    }
+    default fn thunk_serialize(
+        &self,
+        serializer: &mut AnnotatedSerializer,
+    ) -> Result<Document, Error> {
+        self.serialize(serializer)
+    }
+}
 
-pub struct AnnotateType {
-    pub id: IdFn,
-    pub cast: CastFn,
+// We use a private trait to identify whether the Serializer passed to
+// serde::Serialize for dyn Annotate is AnnotatedSerializer.
+trait IsAnnotatedSerializer {
```

Comment:
`IsAnnotatedSerializer` should be an `unsafe trait`, as an incorrect implementation will result in UB.

Reply:
Done.